### PR TITLE
Add QC example and extend DDS command UI to use SDK

### LINKED ISSUE #188: fairmq-dds-command-ui: Provide alternative to -n option
Inspect the topo xml spec and calculate the `-n` value.

### PR DESCRIPTION
**Pull request by rbx**
- [x] Add QC example that demonstrates how to use a config value to change the behaviour of running state handler.
- [x] Add WaitForState to SDK. 
- [x] Refactor fairmq-dds-command-ui to use SDK functionality.
- [x] SDK should unsubscribe from state changes for devices that did not yet exit when SDK is destructed.

Fixes #188.

### CODE REVIEW

File: fairmq/plugins/DDS/runDDSCommandUInew.cxx [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    while (true) {
+        if (command == "c") {
+            cout << "> checking state of the devices" << endl;
+            topo.GetCurrentState();
+            // TODO: extend me
+        } else if (command == "o") {
+            cout << "> dumping config of the devices" << endl;
+            auto const result = topo.GetProperties("^(session|id)$", topologyPath);
```
Comment (dennisklein) [inline fairmq/plugins/DDS/runDDSCommandUInew.cxx]: Codacy Issue found: [Variable 'result' is assigned a value that is never used.](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- fairmq/sdk/Topology.h  6
- examples/qc/runSink.cxx  2
- examples/qc/runSampler.cxx  2
- fairmq/plugins/DDS/runDDSCommandUInew.cxx  5
- test/sdk/_topology.cxx  4
         

Clones removed
==============
+ fairmq/sdk/commands/Commands.cxx  -3
+ fairmq/sdk/commands/Commands.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

File: fairmq/plugins/DDS/runDDSCommandUInew.cxx [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            topo.ChangeState(TopologyTransition::ResetDevice);
+        } else if (command == "h") {
+            cout << "> help" << endl;
+            printControlsHelp();
+        } else if (command == "q") {
+            cout << "> end" << endl;
+            topo.ChangeState(TopologyTransition::End);
+            // TODO: extend me..?
+        } else {
+            cout << "\033[01;32mInvalid input: [" << c << "]\033[0m" << endl;
```
Comment (dennisklein) [inline fairmq/plugins/DDS/runDDSCommandUInew.cxx]: Codacy Issue found: [Uninitialized variable: c](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- fairmq/sdk/Topology.h  6
- test/sdk/_topology.cxx  3
- examples/qc/runSampler.cxx  2
- fairmq/plugins/DDS/runDDSCommandUInew.cxx  6
- examples/qc/runSink.cxx  2
         

Clones removed
==============
+ fairmq/sdk/commands/Commands.cxx  -3
+ fairmq/sdk/commands/Commands.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

File: fairmq/plugins/DDS/runDDSCommandUInew.cxx (hunk 1 above)
Comment (dennisklein) [inline fairmq/plugins/DDS/runDDSCommandUInew.cxx]: Codacy Issue found: [Variable 'result' is assigned a value that is never used.](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642)

File: fairmq/plugins/DDS/runDDSCommandUInew.cxx (hunk 2 above)
Comment (dennisklein) [inline fairmq/plugins/DDS/runDDSCommandUInew.cxx]: Codacy Issue found: [Uninitialized variable: c](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- examples/qc/runSink.cxx  2
- fairmq/plugins/DDS/runDDSCommandUInew.cxx  4
- test/sdk/_topology.cxx  3
- fairmq/sdk/Topology.h  6
- examples/qc/runSampler.cxx  2
         

Clones removed
==============
+ fairmq/sdk/commands/Commands.h  -3
+ fairmq/sdk/commands/Commands.cxx  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

File: fairmq/plugins/DDS/runDDSCommandUInew.cxx [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            topo.ChangeState(TopologyTransition::ResetDevice, std::chrono::milliseconds(timeout));
+        } else if (command == "h") {
+            cout << "> help" << endl;
+            printControlsHelp();
+        } else if (command == "q") {
+            cout << "> end" << endl;
+            topo.ChangeState(TopologyTransition::End, std::chrono::milliseconds(timeout));
+            // TODO: extend me..?
+        } else {
+            cout << "\033[01;32mInvalid input: [" << c << "]\033[0m" << endl;
```
Comment (dennisklein) [inline fairmq/plugins/DDS/runDDSCommandUInew.cxx]: Codacy Issue found: [Uninitialized variable: c](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642)

File: fairmq/plugins/DDS/runDDSCommandUInew.cxx [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    while (true) {
+        if (command == "c") {
+            cout << "> checking state of the devices" << endl;
+            topo.GetCurrentState();
+            // TODO: extend me
+        } else if (command == "o") {
+            cout << "> dumping config of the devices" << endl;
+            auto const result = topo.GetProperties("^(session|id)$", path, std::chrono::milliseconds(timeout));
```
Comment (dennisklein) [inline fairmq/plugins/DDS/runDDSCommandUInew.cxx]: Codacy Issue found: [Variable 'result' is assigned a value that is never used.](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Clones added
============
- examples/qc/runSampler.cxx  2
- examples/qc/runSink.cxx  2
- test/sdk/_topology.cxx  2
- fairmq/sdk/Topology.h  6
         

Clones removed
==============
+ fairmq/plugins/DDS/runDDSCommandUI.cxx  -2
+ fairmq/sdk/commands/Commands.cxx  -3
+ fairmq/sdk/commands/Commands.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

File: fairmq/plugins/DDS/runDDSCommandUI.cxx [hunk 5]
```diff
@@ -49,279 +49,155 @@ struct TerminalConfig
...
-    TerminalConfig tconfig;
-
-    if (command == "") {
+    if (command == "c") {
+        cout << "> checking state of the devices" << endl;
+        topo.GetCurrentState();
+        // TODO: extend me
+    } else if (command == "o") {
+        cout << "> dumping config of the devices" << endl;
+        auto const result = topo.GetProperties("^(session|id)$", path, std::chrono::milliseconds(timeout));
```
Comment (dennisklein) [inline fairmq/plugins/DDS/runDDSCommandUI.cxx]: Codacy Issue found: [Variable 'result' is assigned a value that is never used.](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- examples/qc/runSampler.cxx  2
- fairmq/sdk/Topology.h  6
- test/sdk/_topology.cxx  2
- examples/qc/runSink.cxx  2
         

Clones removed
==============
+ fairmq/plugins/DDS/runDDSCommandUI.cxx  -2
+ fairmq/sdk/commands/Commands.cxx  -3
+ fairmq/sdk/commands/Commands.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

**Comment by rbx** [comment]
Today I observed several times that the tests for the DDS example and also for the new QC example in this PR sometimes hang during the end transition. Not yet sure which of my changes triggered this. Hopefully it is the same issue as DDS team reported, meaning we can reproduce and fix it soon.
Other SDK tests never showed this issue so far.

File: examples/dds/fairmq-start-ex-dds.sh.in [hunk 6]
```diff
@@ -25,57 +25,56 @@ plugin=${1:-localhost}
 
 exec 5>&1
 output=$(dds-session start | tee >(cat - >&5))
-export DDS_SESSION_ID=$(echo ${output} | grep "DDS session ID: " | cut -d' ' -f4)
-echo "SESSION ID: ${DDS_SESSION_ID}"
+export FAIRMQ_DDS_SESSION_ID=$(echo ${output} | grep "DDS session ID: " | cut -d' ' -f4)
+echo "SESSION ID: ${FAIRMQ_DDS_SESSION_ID}"
```
Comment (dennisklein) [inline examples/dds/fairmq-start-ex-dds.sh.in]: See https://github.com/FairRootGroup/DDS/commit/2814ff33af62db684692542899692720ffdf6a8f.

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- examples/qc/runSink.cxx  2
- test/sdk/_topology.cxx  2
- fairmq/sdk/Topology.h  6
- examples/qc/runSampler.cxx  2
         

Clones removed
==============
+ fairmq/sdk/commands/Commands.h  -5
+ fairmq/sdk/commands/Commands.cxx  -5
+ fairmq/plugins/DDS/runDDSCommandUI.cxx  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

File: fairmq/sdk/Topology.h [hunk 7]
```diff
@@ -250,11 +251,18 @@ class BasicTopology : public AsioBase<Executor, Allocator>
     BasicTopology(BasicTopology&&) = default;
     BasicTopology& operator=(BasicTopology&&) = default;
 
-    ~BasicTopology()
+    void UnsubscribeFromStateChanges()
     {
-        
+        using namespace fair::mq::sdk::cmd;
 
+        fDDSSession.SendCommand(Cmds(make<UnsubscribeFromStateChange>()).Serialize());
```
Comment (dennisklein) [inline fairmq/sdk/Topology.h]: Thinking about this: Currently, there is a n-to-1 relationship between `sdk::Topology` objects and `sdk::DDSSession::Impl` objects. In the current implementation (also already before this commit) a second instantiation of a `sdk::Topology` for the same DDS session will result in two state change subscriptions from the same DDS channel ID which then will be folded by DDS to one really, which is probably fine for subscribe, but the first unsubscribe will also unsubscribe all other `sdk::Topology`s. Also, not sure, what happens, if the last custom command callback overrides previous ones? So, we should probably improve here at some point and add a (de)multiplexer to `sdk::DDSSession`.

Comment (dennisklein) [reply fairmq/sdk/Topology.h]: (Out of scope of this PR)

File: examples/dds/fairmq-start-ex-dds.sh.in (hunk 6 above)
Comment (rbx) [reply examples/dds/fairmq-start-ex-dds.sh.in]: Reverted it back to the old name.

**Comment by rbx** [comment]
Fixed the SDK hanging at the End transition. The timeout mechanism was initiated too early during the plugin lifetime, failing to provide necessary time for DDS to send out the final Exiting message.

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- examples/qc/runSink.cxx  2
- examples/qc/runSampler.cxx  2
- test/sdk/_topology.cxx  2
- fairmq/sdk/Topology.h  6
         

Clones removed
==============
+ fairmq/sdk/commands/Commands.cxx  -5
+ fairmq/sdk/commands/Commands.h  -5
+ fairmq/plugins/DDS/runDDSCommandUI.cxx  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- test/sdk/_topology.cxx  2
- fairmq/sdk/Topology.h  6
         

Clones removed
==============
+ fairmq/plugins/DDS/runDDSCommandUI.cxx  -2
+ fairmq/sdk/commands/Commands.cxx  -5
+ fairmq/sdk/commands/Commands.h  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- fairmq/sdk/Topology.h  6
- test/sdk/_topology.cxx  2
         

Clones removed
==============
+ fairmq/plugins/DDS/runDDSCommandUI.cxx  -2
+ fairmq/sdk/commands/Commands.cxx  -5
+ fairmq/sdk/commands/Commands.h  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- test/sdk/_topology.cxx  2
- fairmq/sdk/Topology.h  6
         

Clones removed
==============
+ fairmq/sdk/commands/Commands.h  -5
+ fairmq/sdk/commands/Commands.cxx  -5
+ fairmq/plugins/DDS/runDDSCommandUI.cxx  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5000642&bid=16500701)

**Pull request merged**